File: src/test/java/testingil/webinar/tddapi/s13/pass_unit_test/NavigatorUnitTests.java
package testingil.webinar.tddapi.s13.pass_unit_test;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import testingil.webinar.tddapi.s13.Distance;
import testingil.webinar.tddapi.s13.Location;
import testingil.webinar.tddapi.s13.Navigator;
import testingil.webinar.tddapi.s13.DistanceProvider;

class NavigatorUnitTests {

	@Test
	void when_created_distance_is_always_zero() {
		Location loc = new Location("London");
		DistanceProvider mockDistService = 
				mock(DistanceProvider.class, RETURNS_DEEP_STUBS);
		Navigator nav = new Navigator(loc, mockDistService);
		Distance distance = nav.getDistanceFromDestination();
		assertThat(distance.inKm(), is(0));
	}
	
	@Test
	@DisplayName("2500 miles = 4000 kms")
	void distance_is_calculated_in_km() {
		DistanceProvider mockDistService = 
				mock(DistanceProvider.class);

		Distance distanceInMiles = new Distance();
		distanceInMiles.value = 2500;
		when(mockDistService.getDistance(
				Mockito.any(Location.class), 
				Mockito.any(Location.class)))
				.thenReturn(distanceInMiles);
		
		Location startingPoint = new Location("New York City");
		Navigator nav = new Navigator(startingPoint, mockDistService);
		Location dest = new Location("Los Angeles");
		nav.setDestination(dest);
		Distance distance = nav.getDistanceFromDestination();
		assertThat(distance.inKm(), is(4000));
	}

}

File: Python/s13_pass_unit_test/navigator_unit_tests.py
import unittest
from unittest.mock import patch

from hamcrest import assert_that, equal_to

from s13_pass_unit_test.distance import Distance
from s13_pass_unit_test.location import Location
from s13_pass_unit_test.navigator import Navigator


class NavigatorUnitTests(unittest.TestCase):
    def test_when_created_distance_is_always_zero(self):
        distProvider = patch('s11_mock_adapter.distance_adapter.DistanceAdapter').start()
        distProvider.get_distance.return_value = Distance(0)

        loc = Location("London")
        nav = Navigator(loc, distProvider)
        distance = nav.get_distance_from_destination()
        assert_that(distance.inKm(), equal_to(0))

    def test_distance_is_calculated_in_km(self):
        """2500 miles = 4000 kms"""
        distProvider = patch('s11_mock_adapter.distance_adapter.DistanceAdapter').start()
        distProvider.get_distance.return_value = Distance(2500)
        startingPoint = Location("New York City")
        nav = Navigator(startingPoint, distProvider)
        dest = Location("Los Angeles")
        nav.set_destination(dest)
        distance = nav.get_distance_from_destination()
        assert_that(distance.inKm(), equal_to(4000))



if __name__ == '__main__':
    unittest.main()
